File: nvpl_blas/c/zher.c
/******************************************************************************
 * Content: 
 *     This example demonstrates use of API as below:
 *     cblas_zher
 *
 ******************************************************************************/
#include "example_helper.h"

int main() {
    nvpl_int_t N = 2;
    nvpl_int_t lda = N;
    nvpl_int_t incX = 1;
    double alpha = 1.0f;
    enum CBLAS_ORDER order = CblasRowMajor;
    enum CBLAS_UPLO uplo = CblasUpper;
    nvpl_dcomplex_t * A = NULL;
    nvpl_dcomplex_t * X = NULL;
    nvpl_int_t len_x = 1 + (N - 1) * labs(incX);

    printf("\nExample: cblas_zher for the Hermitian rank-1 update\n\n");
    printf("#### args: n=%" PRId64 ", lda=%" PRId64 ", incX=%" PRId64 ", order=%c, uplo=%c\n",
           (int64_t)N, (int64_t)lda, (int64_t)incX, order_to_char(order), uplo_to_char(uplo));
    printf("           alpha=%g\n", alpha);

    // allocate memory
    A = (nvpl_dcomplex_t *)malloc(lda * N  * sizeof(nvpl_dcomplex_t));
    X = (nvpl_dcomplex_t *)malloc(len_x * sizeof(nvpl_dcomplex_t));

    // fill data
    fill_zmatrix(A, N, N, lda, order, UPLO_TO_FILL_MODE(uplo), CblasNonUnit);
    fill_zvector(X, N, incX);

    // print input data
    print_zmatrix(A, N, N, lda, "A", order);
    print_zvector(X, N, incX, "X");

    // call cblas_zher
    cblas_zher(order, uplo, N, alpha, X, incX, A, lda);

    // print result
    printf("\nAfter zher operation:\n");
    print_zmatrix(A, N, N, lda, "A", order);

    // release memory
    free(A);
    free(X);
    return 0;
}

File: nvpl_blas/f77/zher.f
!******************************************************************************
! Content:
!     This example demonstrates use of API as below:
!     ZHER 
!
!******************************************************************************
      program   ZHER_MAIN
*
      implicit none

      include "nvpl_blas_f77_blas.fi"

*       Intrinsic Functions
      intrinsic  abs

      integer         n, incx, i, lda
      parameter       (n=2, incx=1, lda=2)
      double precision alpha
      parameter       (alpha=1.0)
      integer         xsize
      parameter       (xsize=1+(n-1)*abs(incx))
      double complex  x(xsize), a(lda,n)

*       External Subroutines
      external        print_zvector, fill_zvector, fill_zmatrix,
     &                print_zmatrix
      character*1     uplo
      parameter       (uplo='U')
*
*      Executable Statements
*
      print*
      print 99
*
*       Print input data
      print*
      print 100, n, lda, incx
      print 101, alpha
      call fill_zvector(x,n,incx)
      call fill_zmatrix(a,n,n,lda,uplo,'N')
      call print_zmatrix(a,n,n,lda,'A')
      call print_zvector(x,n,incx,'X')
*
*      Call ZHER subroutine
      call ZHER(uplo,n,alpha,x,incx,a,lda)
*
      print*
      print 102
      call print_zmatrix(a,n,n,lda,'A')

      stop
 99   format('Example: ZHER for the Hermitian rank-1 update')
 100  format('#### args: n=',i1,', lda=',i1,', incx=',i1)
 101  format(11x,'alpha=',f4.2)
 102  format('The result of ZHER: ')
      end
